turbo/jsonrpc/erigon_receipts.go

```diff
@@ -295,91 +208,96 @@ func (api *ErigonImpl) GetLatestLogs(ctx context.Context, crit filters.FilterCri
-	for iter.HasNext() {
+	it := rawdbv3.TxNums2BlockNums(tx, txNumbers, order.Asc)
+	defer it.Close()
+
+	var blockHash common.Hash
+	var header *types.Header
+	var logCount, blockCount, timestamp uint64
+
+	for it.HasNext() {
 		if err = ctx.Err(); err != nil {
 			return nil, err
 		}
-
-		blockNumber := uint64(iter.Next())
-		var logIndex uint
-		var txIndex uint
-		var blockLogs []*types.Log
-		it, err := tx.Prefix(kv.Log, hexutility.EncodeTs(blockNumber))
+		txNum, blockNum, txIndex, isFinalTxn, blockNumChanged, err := it.Next()
 		if err != nil {
 			return nil, err
 		}
-		for it.HasNext() {
-			k, v, err := it.Next()
-			if err != nil {
-				return erigonLogs, err
-			}
-			var logs types.Logs
-			if err := cbor.Unmarshal(&logs, bytes.NewReader(v)); err != nil {
-				return erigonLogs, fmt.Errorf("receipt unmarshal failed:  %w", err)
-			}
-			for _, log := range logs {
-				log.Index = logIndex
-				logIndex++
-			}
-			var filtered types.Logs
-			var maxLogCount uint64
-			maxLogCount = 0
-			if logOptions.LogCount != 0 {
-				maxLogCount = logOptions.LogCount - logCount
-			}
+		if isFinalTxn {
+			continue
+		}
 
-			if logOptions.IgnoreTopicsOrder {
-				filtered = logs.CointainTopics(addrMap, topicsMap, maxLogCount)
-			} else {
-				filtered = logs.Filter(addrMap, crit.Topics, maxLogCount)
+		// if block number changed, calculate all related field
+		if blockNumChanged {
+			if header, err = api._blockReader.HeaderByNumber(ctx, tx, blockNum); err != nil {
+				return nil, err
 			}
-			if len(filtered) == 0 {
+			if header == nil {
+				log.Warn("[rpc] header is nil", "blockNum", blockNum)
 				continue
 			}
-			txIndex = uint(binary.BigEndian.Uint32(k[8:]))
-			for i := range filtered {
-				filtered[i].TxIndex = txIndex
-			}
-			for i := len(filtered) - 1; i >= 0; i-- {
-				blockLogs = append(blockLogs, filtered[i])
-				logCount++
-			}
-			if logOptions.LogCount != 0 && logOptions.LogCount <= logCount {
-				break
-			}
+			blockHash = header.Hash()
+			exec.ChangeBlock(header)
+			timestamp = header.Time
 		}
-		it.Close()
-		blockCount++
-		if len(blockLogs) == 0 {
+		var logIndex uint
+		var blockLogs types.Logs
+
+		txn, err := api._txnReader.TxnByIdxInBlock(ctx, tx, blockNum, txIndex)
+		if err != nil {
+			return nil, err
+		}
+		if txn == nil {
 			continue
 		}
 
-		header, err := api._blockReader.HeaderByNumber(ctx, tx, blockNumber)
+		_, err = exec.ExecTxn(txNum, txIndex, txn)
 		if err != nil {
 			return nil, err
 		}
-		if header == nil {
-			return nil, fmt.Errorf("block header not found: %d", blockNumber)
+		blockLogs = exec.GetLogs(txIndex, txn)
+		for _, log := range blockLogs {
+			log.Index = logIndex
+			logIndex++
+		}
+		var filtered types.Logs
+		var maxLogCount uint64
+		maxLogCount = 0
+		if logOptions.LogCount != 0 {
+			maxLogCount = logOptions.LogCount - logCount
+		}
+		if logOptions.IgnoreTopicsOrder {
+			filtered = blockLogs.CointainTopics(addrMap, topicsMap, maxLogCount)
```

Comment:
create separated github issue - to investigate: can we remove this checks or not. because in E3 indices are txn-level-granularity - it means all filtering already done by `applyFiltersV3(tx, begin, end, crit)`